compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp

```diff
@@ -59,6 +63,110 @@ static SmallVector<int64_t> getIndicesOfDynamicDims(ShapedType t) {
 
+/// Iteratively populate the computations that define `foreachThreadOp`'s
+/// getNumThreads into the workgroup_count region of `dispatchOp` and set the
+/// `dispatchOp` workload operands to the values required above.
+/// Assumes the Flow::DispatchWorkgroupsOp is built with an empty region.
+static void populateWorkgroupCountComputingRegion(
+    PatternRewriter &rewriter, scf::ForeachThreadOp foreachThreadOp,
+    Flow::DispatchWorkgroupsOp dispatchOp) {
+  Location loc = foreachThreadOp.getLoc();
+  OpBuilder::InsertionGuard g(rewriter);
+  Region &r = dispatchOp.workgroup_count();
+  assert(r.empty() && "expected block-less workgroup_count region");
+  Block *block = rewriter.createBlock(&r);
+  rewriter.setInsertionPointToStart(block);
+  SmallVector<Value> workgroupCount{foreachThreadOp.getNumThreads()};
+  // Resize to `3` to match IREE's assumptions.
+  workgroupCount.resize(3, rewriter.create<arith::ConstantIndexOp>(loc, 1));
+  auto returnOp = rewriter.create<Flow::ReturnOp>(loc, workgroupCount);
+
+  // Start filling the region.
+  // Create a worklist of the Operations creating the values returned.
+  // Iteratively pull the first operation that is not dominated by any others
+  // and add the operations defining it to the worklist unless either:
+  //   1. the operation consumes any tensor, we avoid pulling tensor arguments
+  //      into the workgroup count region.
+  //   2. the operation is null (i.e. the operand is a block argument), in which
+  //      case we just pass the block argument without trying to analyze
+  //      further.
+  //   3. the operation has already been seen, in which case we avoid
+  //      reprocessing.
+  {
+    BlockAndValueMapping bvm;
+    auto parentFuncOp = foreachThreadOp->getParentOfType<func::FuncOp>();
+    DominanceInfo domInfo(parentFuncOp);
+    llvm::SmallPtrSet<Operation *, 16> worklist;
+    for (auto v : foreachThreadOp.getNumThreads()) {
+      if (auto *op = v.getDefiningOp()) {
+        LLVM_DEBUG(DBGS() << "Initial WL op: " << *op << "\n");
+        worklist.insert(op);
+      }
+    }
+
+    // Keep a set of seen ops to avoid reprocessing.
+    llvm::SmallPtrSet<Operation *, 16> seen = worklist;
+    while (!worklist.empty()) {
+      auto nextOpIter = llvm::find_if(worklist, [&](Operation *op1) {
+        LLVM_DEBUG(DBGS() << "Consider WL op: " << *op1 << "\n");
+        // Bail as soon as we see a non-IndexType producing op.
+        // TODO: do we want to relax this and only forbid TensorType/ShapedType?
+        if (llvm::any_of(op1->getOperandTypes(),
+                         [](Type t) { return !t.isa<IndexType>(); }))
+          return false;
+        bool res = llvm::all_of(worklist, [&](Operation *op2) {
+          LLVM_DEBUG(DBGS()
+                     << "\tProperly dominated by op: " << *op2 << " ? -> "
+                     << domInfo.properlyDominates(op2, op1) << "\n");
+          return !domInfo.properlyDominates(op2, op1);
```

Comment:
I am guessing you are just pulling in `arith.constant` statements for now. My suggestion would be to generalize when needed. Start with just pulling in `arith.constant` since thats what you need here. Then if needed we can generalize.

Reply:
Good point about dominance, thanks!
Restricting the algorithm to just pull in constants, I can revive the WL-based algorithm later then.